File: src/35_CopyComplexList/ComplexList.cpp
#include "ComplexList.h"

#include <iostream>
#include <ostream>

Node* CreateNode( int value ) {
    Node* pNode = new Node( value );
    return pNode;
}

void BuildNodes( Node* pNode, Node* pNext, Node* pRand ) {
    if ( pNode == nullptr ) {
        std::cout << "Error to connect nodes.\n";
        exit( 1 );
    }
    pNode->next = pNext;
    pNode->random = pRand;
}

void PrintList( Node* pHead ) {
    Node* pNode = pHead;
    while ( pNode != nullptr ) {
        std::cout << "The value of this node is: " << pNode->val << std::endl;
        if ( pNode->random != nullptr )
            std::cout << "The value of its random sibling is: "
                      << pNode->random->val << std::endl;
        else
            std::cout << "This node does not have a sibling.\n";
        std::cout << std::endl;
        pNode = pNode->next;
    }
}

void DestroyList( Node* pHead ) {
    Node* pNode = pHead;
    while ( pNode != nullptr ) {
        pHead = pHead->next;
        delete pNode;
        pNode = pHead;
    }
}

File: src/35_CopyComplexList/complex_list.py
class Node:
    def __init__(self, x: int, next: "Node" = None, random: "Node" = None):
        self.val = int(x)
        self.next = next
        self.random = random


def CreateNode(value) -> Node:
    return Node(value)


def BuildNodes(curNode: Node, nextNode: Node, randNode: Node):
    if curNode is None:
        print("Error to connect two nodes.")
    curNode.next = nextNode
    curNode.random = randNode


def PrintList(head: Node):
    node = head
    while node is not None:
        print(f"The value of this node is: {node.val}")
        if node.random is not None:
            print(f"The value of its random sibling is: {node.random.val}\n")
        else:
            print("This node does not have a sibling.\n")
        node = node.next
